gossip/src/crds.rs

```diff
@@ -147,14 +149,20 @@ impl Cursor {
 }
 
 impl VersionedCrdsValue {
-    fn new(value: CrdsValue, cursor: Cursor, local_timestamp: u64) -> Self {
+    fn new(value: CrdsValue, cursor: Cursor, local_timestamp: u64, route: GossipRoute) -> Self {
         let value_hash = hash(&serialize(&value).unwrap());
+        let num_push_recv = match route {
+            // If we received a PushMessage, set
+            // num_push_recv <- 1
+            GossipRoute::PushMessage(_) => 1u8,
+            _ => 0u8,
```

Comment:
do we have to pay attention to `LocalMessage` or `PullRequest` here? At the very least I think we can ignore `PullRequest` since we're not updating our `crds.table` when we receive a `PullRequest`

Reply:
> we're not updating our crds.table when we receive a PullRequest

We actually upsert the origin of `PullRequest` into crds table:
https://github.com/anza-xyz/agave/blob/0f1ca20d3/gossip/src/crds_gossip_pull.rs#L296

My initial thought was that we can just leave it zero anyways, but looks like it might break `record_redundant_pull` below.

An ugly workaround would be define:
```rust
num_push_recv: Option<u8>,
```

Then in `VersionedCrdsValue::new`:
```rust
let num_push_recv = match route {
   GossipRoute::LocalMessage => None,
   GossipRoute::PullRequest => None,
   GossipRoute::PullResponse => Some(0),
   GossipRoute::PushMessage => Some(1),
}
```
Then `Some(0)` will accurately identify redundant pulls.

Reply:
ahh shoot missed that. ya not an ideal option but would work. i'll think...